Bottom: f46e3dd8dfe1f6b1863cabb7c6d91898fb376f83
Top:    2a7c9bb5600112f68a6f3f8ed162e6f985ca8a4b
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2010-12-09 23:08:11 +0000

WORK IN PROGRESS


---

diff --git a/src/hid/common/hidnogui.c b/src/hid/common/hidnogui.c
index 2b98a75..18bc637 100644
--- a/src/hid/common/hidnogui.c
+++ b/src/hid/common/hidnogui.c
@@ -294,14 +294,23 @@ static char *
 nogui_prompt_for (const char *msg, const char *default_string)
 {
   static char buf[1024];
+  char *s;
+
   if (default_string)
     printf ("%s [%s] : ", msg, default_string);
   else
     printf ("%s : ", msg);
-  fgets (buf, 1024, stdin);
-  if (buf[0] == 0 && default_string)
-    strcpy (buf, default_string);
-  return buf;
+
+  s = fgets (buf, 1024, stdin);
+  if (s == NULL || buf[0] == '\0' || buf[0] == '\r' || buf[0] == '\n')
+    {
+      if (default_string != NULL)
+        return strdup (default_string);
+      else
+        return strdup ("");
+    }
+  else
+    return strdup (buf);
 }
 
 /* FIXME - this could use some enhancement to actually use the other
@@ -312,14 +321,23 @@ nogui_fileselect (const char *title, const char *descr,
 		  const char *history_tag, int flags)
 {
   static char buf[1024];
+  char *s;
+
   if (default_file)
     printf ("%s [%s] : ", title, default_file);
   else
     printf ("%s : ", title);
-  fgets (buf, 1024, stdin);
-  if (buf[0] == 0 && default_file)
-    strcpy (buf, default_file);
-  return buf;
+
+  s = fgets (buf, 1024, stdin);
+  if (s == NULL || buf[0] == '\0' || buf[0] == '\r' || buf[0] == '\n')
+    {
+      if (default_file != NULL)
+        return strdup (default_file);
+      else
+        return NULL;
+    }
+  else
+    return strdup (buf);
 }
 
 static int
